Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+                key: “is_another_feature_enabled”
+            )
+        }
+        
+        // static configurations
+        public let config: Configuration
+        
+        public struct Configuration {
+            public let someAppSpecificConfiguration: Bool // can have a default value
+        }
+    }
+}
+```
+
+This way feature switches will be accessed in the feature builders via `Current.someFeature.featureSwitches` or `Current.someFeature.config`.
+
+If the type of the switch changes (i.e. from local to remote switch) we change it in a single place where it is defined rather in every place where it is used. To abstract away difference between static configuration and dynamic switches we can use a key-path based getters or subscripts:
+
+```swift
+extension World.SomeFeatureModule {
+    public func get<T>(_ keyPath: KeyPath<RemoteFeatureSwitch<T>, World.SomeFeatureModule.FeatureSwitches>) -> T {
+        return featureSwitches.keyPath[keyPath].value
+    }
+    
+    public func get<T>(_ keyPath: KeyPath<T, World.SomeFeatureModule.Configuration>) -> T {
+        return config.keyPath[keyPath]
+    }
+}
+```
+
+With Swift 5.1 we can improve this a bit with property wrappers which will allow us to use `isNewFeatureEnabled` directly rather than with `isNewFeatureEnabled.value`:
+
+```swift
+@RemoteFeatureSwitch(
+    name: “New feature”,
+    key: “is_new_feature_enabled”
+)
+var isNewFeatureEnabled: Bool
+```
+
+Also we will be able to use `dynamicMemberLookup` with keypaths instead of custom getters or subscripts so that we wil be able to access dynamic feature switches or static configuration as `Current.someFeature.isNewFeatureEnabled` instead of `Current.someFeatures.get(\.isNewFeatureEnabled)`
+
+### "Сoncretize" app configurations
+
+We already moved `AppConfigurationProtocol` to `BabylonDependencies` so that the instance of app configuration is a part of `Current`. But that does not solve all the issues with amount of boilerplate we have to write to add a new configuration and discoverability of values.
+
+To solve these issues we propose to replace `protocol AppConfigurationProtocol` with concrete `struct AppConfiguration` defined in `BabylonDependencies` and initialise it with application specific values in a target specific code. This will reduce amount of boilerplate to write, but will still require us to change code multiple times in each target when something changes in this struct. I.e. when we add a new feature switch to `AppConfiguration` we will need to add it to the paramters that we pass to it's initializer in each target. We can use default values where possible to avoid that. For that we should agree that default values are always `false` so that we can definetely know what is the value if it is not passed to the constructor. Protocol and default protocol implementations achive the same result but with much more boilerplate (remember point-free issue about using value types instead of protocols)
```

Comment:
```suggestion
To solve these issues we propose to replace `protocol AppConfigurationProtocol` with concrete `struct AppConfiguration` defined in `BabylonDependencies` and initialise it with application specific values in a target specific code. This will reduce amount of boilerplate to write, but will still require us to change code multiple times in each target when something changes in this struct. I.e. when we add a new feature switch to `AppConfiguration` we will need to add it to the parameters that we pass to it's initializer in each target. We can use default values where possible to avoid that. For that we should agree that default values are always `false` so that we can definetely know what is the value if it is not passed to the constructor. Protocol and default protocol implementations achive the same result but with much more boilerplate (remember point-free issue about using value types instead of protocols)
```